tinygrad/runtime/ops_clang.py

```diff
@@ -1,21 +1,45 @@
 from typing import Optional, List
-import ctypes, subprocess, pathlib, tempfile
+import platform, struct, ctypes, subprocess, pathlib, tempfile, tinygrad.runtime.autogen.libc as libc
+from mmap import PROT_READ, PROT_WRITE, PROT_EXEC, MAP_ANON, MAP_PRIVATE
+from tinygrad.runtime.support.elf import elf_loader
 from tinygrad.device import Compiled, Compiler, MallocAllocator
-from tinygrad.helpers import cpu_time_execution, DEBUG, cpu_objdump
+from tinygrad.helpers import cpu_time_execution, DEBUG, OSX, getenv, cpu_objdump
 from tinygrad.renderer.cstyle import ClangRenderer
 
+MAP_JIT, global_handle = 0x0800 if OSX else 0, ctypes.CDLL(None)
+def i2u(bits:int, x:int): return 2**bits+x if x < 0 else x
+def patchuint32(image: memoryview, ploc: int, new: int): image[ploc:ploc+4] = struct.pack("<I", struct.unpack("<I", image[ploc:ploc+4])[0] | new)
+SIMPLE_AARCH64_RELOCATIONS = [libc.R_AARCH64_ADD_ABS_LO12_NC, libc.R_AARCH64_LDST16_ABS_LO12_NC, libc.R_AARCH64_LDST32_ABS_LO12_NC,
+                              libc.R_AARCH64_LDST64_ABS_LO12_NC, libc.R_AARCH64_LDST128_ABS_LO12_NC]
+
 class ClangCompiler(Compiler):
   def __init__(self, cachekey="compile_clang", args:Optional[List[str]]=None):
-    self.args = ['-march=native'] if args is None else args
+    self.args = ['-shared', '-march=native'] if args is None else args
     super().__init__(cachekey)
 
   def compile(self, src:str) -> bytes:
     # TODO: remove file write. sadly clang doesn't like the use of /dev/stdout here
     with tempfile.NamedTemporaryFile(delete=True) as output_file:
-      subprocess.check_output(['clang', '-shared', *self.args, '-O2', '-Wall', '-Werror', '-x', 'c', '-fPIC', '-ffreestanding', '-nostdlib',
+      subprocess.check_output(['clang', *self.args, '-O2', '-Wall', '-Werror', '-x', 'c', '-fPIC', '-ffreestanding', '-nostdlib',
                                '-', '-o', str(output_file.name)], input=src.encode('utf-8'))
       return pathlib.Path(output_file.name).read_bytes()
 
+class ClangJITCompiler(ClangCompiler):
+  def __init__(self, key="compile_clang_jit"):
+    # x18 is reserved register on at least macOS and Windows and is clobbered on context switch
+    super().__init__(key, ['-c', '-march=native', f'--target={platform.machine()}-none-unknown-elf', '-fno-math-errno',
```

Comment:
Does this need an output file?

Reply:
The ClangJITCompiler doesn't need it (actually the thing that doesn't like output to stdout is a linker if i recall correctly). I reverted my don't use file thing in v2 because now ClangCompiler is used by some other parts of tinygrad that might require it